pkg/controller/mysqldatabase/db_controller_test.go

```diff
@@ -0,0 +1,315 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	mysqlv1alpha1 "github.com/presslabs/mysql-operator/pkg/apis/mysql/v1alpha1"
+	"github.com/presslabs/mysql-operator/pkg/controller/internal/testutil"
+	"github.com/presslabs/mysql-operator/pkg/internal/mysql/fake"
+	"github.com/presslabs/mysql-operator/pkg/internal/mysqldatabase"
+	"github.com/presslabs/mysql-operator/pkg/testutil/factories"
+	gm "github.com/presslabs/mysql-operator/pkg/testutil/gomegamatcher"
+)
+
+var _ = Describe("MySQL database controller", func() {
+	var (
+		// channel for incoming reconcile requests
+		requests chan reconcile.Request
+		// stop channel for controller manager
+		stop chan struct{}
+
+		// controller k8s client
+		c client.Client
+
+		fakeQR *fake.QueryRunner
+	)
+
+	BeforeEach(func() {
+		mgr, err := manager.New(cfg, manager.Options{
+			Scheme:             scheme.Scheme,
+			MetricsBindAddress: "0",
+		})
+		Expect(err).NotTo(HaveOccurred())
+
+		// create new k8s client
+		// NOTE: create a new k8s client without cache to have more stable tests
+		c, err = client.New(cfg, client.Options{})
+		Expect(err).To(Succeed())
+
+		fakeQR = fake.NewQueryRunner(false)
+
+		var recFn reconcile.Reconciler
+		recFn, requests = testutil.SetupTestReconcile(newReconciler(mgr, fakeQR.Run))
+		Expect(add(mgr, recFn)).To(Succeed())
+
+		stop = testutil.StartTestManager(mgr)
+	})
+
+	AfterEach(func() {
+		close(stop)
+	})
+
+	When("db resource is created", func() {
+		var (
+			clusterName     string
+			db              *mysqldatabase.Database
+			expectedRequest reconcile.Request
+		)
+		BeforeEach(func() {
+			clusterName = fmt.Sprintf("mysql-%d", rand.Int())
+			db = factories.NewDatabase(factories.WithMySQLCluster(context.TODO(), c, clusterName))
+
+			fakeQR.AddExpectedCalls(
+				func(dsn string, query string, args ...interface{}) error {
+					defer GinkgoRecover()
+
+					Expect(dsn).To(Equal(getExpectedDSN(clusterName)))
+					Expect(query).To(Equal(fmt.Sprintf("CREATE DATABASE IF NOT EXISTS `%s` CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci", db.Name)))
+					Expect(args).To(HaveLen(0))
+
+					return nil
+				},
+				func(dsn string, query string, args ...interface{}) error {
+					defer GinkgoRecover()
+
+					return nil
+				},
+			)
+
+			// create DB in k8s
+			Expect(c.Create(context.TODO(), db.Unwrap())).To(Succeed())
+
+			expectedRequest = reconcile.Request{NamespacedName: types.NamespacedName{Name: db.Name, Namespace: db.Namespace}}
+
+			// first event when the resource is created
+			Eventually(requests).Should(Receive(Equal(expectedRequest)))
+
+			// second event when is updated with the finalizer and status
+			Eventually(requests, "2s").Should(Receive(Equal(expectedRequest)))
+
+			// should not receive any new reconcile requests
+			Consistently(requests).ShouldNot(Receive(Equal(expectedRequest)))
+		})
+
+		AfterEach(func() {
+			fakeQR.AssertNoCallsLeft()
+			fakeQR.AllowExtraCalls()
```

Comment:
`AssertNoCallsLeft()` checks if there are any extra calls left in the buffer while `AllowExtraCalls()` allows to make new calls (usually calls made when the resource is deleted because in the `AfterEach` block the resource is deleted)